### ISSUE TITLE
Support for Intel Integrated Graphics Processor cores

### ISSUE THREAD

**Issue by clvgt12** [ollama/ollama#3113]
Hello,

Please consider adapting Ollama to use Intel Integrated Graphics Processors (such as the Intel Iris Xe Graphics cores) in the future.

**Comment by ddpasa** [comment]
take a look at: https://github.com/ollama/ollama/pull/2578

**Comment by clvgt12** [comment]
Very nice!   looking forward to testing it on my Windows PC running Ollama in the future!

**Comment by vinaykharayat** [comment]
+1

**Comment by MarkWard0110** [comment]
I would imagine for anyone who has an Intel integrated GPU, the otherwise unused GPU would add an additional GPU to utilize.  
Even if it was limited to 3GB.  That would be an additional 3GB GPU that could be utilized.  It is a 3GB GPU that is not utilized when a model is split between an Nvidia GPU and CPU.
I am running a headless server and the integrated GPU is there and not doing anything to help.

**Comment by carlos-burelo** [comment]
+1

**Comment by sspanogle** [comment]
+1 xo

**Comment by suncloudsmoon** [comment]
+1

**Comment by serhatsatir** [comment]
I also have Intel Iris Xe with 8GB ram. But I can't see any benefit. It would be very useful if the hardware we have could be used to its full capacity. 
Perhaps consulting AI on how to do it could be a solution. 😂

**Comment by carlos-burelo** [comment]
Personally, I have tried using [WebLLM](https://github.com/mlc-ai/web-llm) to run AI models like Llama3. When I do this, I notice an improvement in token generation speed because the Intel graphics card is being utilized via the WebGPU API. I wouldn't say the improvement is radical, but it is slightly faster. With some caution, I would estimate the improvement to be around 15%. My specifications are:

Chipset: Intel Core i7 1165G7
Graphics: Intel Iris Xe Graphics 15.7 GB shared memory
RAM: 32 GB DDR4 2667MT/s

Therefore, results may vary significantly depending on the specifications of each system.

**Comment by jomardyan** [comment]
+1

**Comment by LuckyMcBeast** [comment]
+1

**Comment by ThatMarko** [comment]
+1

**Comment by ayttop** [comment]
ollama with igpu intel

**Comment by user7z** [comment]
> ollama with igpu intel
> 
> ollama run with igpu intel with
> 
> https://github.com/intel-analytics/ipex-llm/blob/main/docs/mddocs/Quickstart/llama_cpp_quickstart.md

It poorly do , it needs a special setup with a special distrbution , it didnt work with me on a non apt distrbution , see this [issue](https://github.com/intel-analytics/ipex-llm/issues/12120#issuecomment-2379372351)

**Comment by havardthom** [comment]
For anyone interested, I've added an Ollama LXC script to tteck's Proxmox Helper-Scripts. The script installs intel-basekit and builds Ollama from source and supports Intel iGPU passthrough (though it has a very long install time). It can be run like any other proxmox helper script: `bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/ollama.sh)"`

A script for Open WebUI LXC with optional Ollama install is also available: https://tteck.github.io/Proxmox/#open-webui-lxc

**Comment by maczet** [comment]
+1

**Comment by SystemStrategy** [comment]
I ran the tteck ollama and would not load some models and was way slower compared to the docker version of Ipex

Intel Docker IPEX:
https://github.com/intel-analytics/ipex-llm/blob/main/docs/mddocs/DockerGuides/docker_cpp_xpu_quickstart.md

Added commands to auto-start:
https://github.com/SystemStrategy/Proxmox/blob/main/Ipex_Compose

**Comment by gaborkukucska** [comment]
+1

**Comment by ChrisBGL** [comment]
Please add native support for intel iGPU

**Comment by user7z** [comment]
I think its not ollama devs probleme , itd an intel probleme that they cant make their oneAPI be usable by the community , & the obscure way ipex-llm is bieng developped is just insane & wouldnt make it possible for ollama devs to integrate it , its intel's probleme.

**Comment by MaoJianwei** [comment]
Can ollama use Intel integrated GPU to speed up inference？e.g. Intel UHD Graphics 630 of i5-10400

**Comment by ddpasa** [comment]
> Can ollama use Intel integrated GPU to speed up inference？e.g. Intel UHD Graphics 630 of i5-10400

I'm using an intel iGPU with llama.cpp with Vulkan.  Ollama is just a wrapper for llama.cpp, but for reasons unknown the devs refuse to enable Vulkan.  If you have an Intel iGPU, my recommendation is to use llama.cpp directly.

**Comment by NeoZhangJianyu** [comment]
This issue has a solution by https://github.com/ollama/ollama/issues/8414.

**Comment by MaoJianwei** [comment]
> > Can ollama use Intel integrated GPU to speed up inference？e.g. Intel UHD Graphics 630 of i5-10400
> 
> I'm using an intel iGPU with llama.cpp with Vulkan. Ollama is just a wrapper for llama.cpp, but for reasons unknown the devs refuse to enable Vulkan. If you have an Intel iGPU, my recommendation is to use llama.cpp directly.



@ddpasa Many thanks, I have tried to run Ollama with Irix Xe(iGPU) of Intel i7, but I found the inference speed of iGPU is near to that of CPU.
**(about 18 tokens/s against 19 tokens/s)**

So I think it makes no sence to attemp to run Ollama with iGPU.

**Comment by MaoJianwei** [comment]
> This issue has a solution by [#8414](https://github.com/ollama/ollama/issues/8414).

No, #8414 doesn't support 10th Intel CPU @NeoZhangJianyu

**Comment by Gunnarr970** [comment]
> So I think it makes no sence to attemp to run Ollama with iGPU.

Even if the speed is the same, CPU resources will continue to be available for other processes if the GPU is used

**Comment by NeoZhangJianyu** [comment]
> > This issue has a solution by [#8414](https://github.com/ollama/ollama/issues/8414).
> 
> No, [#8414](https://github.com/ollama/ollama/issues/8414) doesn't support 10th Intel CPU [@NeoZhangJianyu](https://github.com/NeoZhangJianyu)

The iGPU in 10th CPU don't supported by oneAPI (SYCL).
It's the root cause that llama.cpp SYCL backend can't support it.

Refer to [hardware](https://github.com/ggml-org/llama.cpp/blob/master/docs/backend/SYCL.md#hardware).

**Comment by MaoJianwei** [comment]
> > So I think it makes no sence to attemp to run Ollama with iGPU.
> 
> Even if the speed is the same, CPU resources will continue to be available for other processes if the GPU is used

Yes, but my purpose of using iGPU is to speed up inference, this expection is not satisfied.

**Comment by MaoJianwei** [comment]
> 
> The iGPU in 10th CPU don't supported by oneAPI (SYCL). It's the root cause that llama.cpp SYCL backend can't support it.
> 
> Refer to [hardware](https://github.com/ggml-org/llama.cpp/blob/master/docs/backend/SYCL.md#hardware).

Thanks, I see. I bought my computer one year earlier.  What a poty.

**Comment by ddpasa** [comment]
> > > So I think it makes no sence to attemp to run Ollama with iGPU.
> > 
> > 
> > Even if the speed is the same, CPU resources will continue to be available for other processes if the GPU is used
> 
> Yes, but my purpose of using iGPU is to speed up inference, this expection is not satisfied.

Token generation is limited by memory bandwidth, so you'll see very similar speeds for CPU or iGPU.  Th iGPU helps with input token processing or image processing.  I'm getting x2 to x3 speedups on input token processing or VLMs on an Intel 10th gen iGPU when using llama.cpp with Vulkan.

**Comment by MaoJianwei** [comment]
> > > > So I think it makes no sence to attemp to run Ollama with iGPU.
> > > 
> > > 
> > > Even if the speed is the same, CPU resources will continue to be available for other processes if the GPU is used
> > 
> > 
> > Yes, but my purpose of using iGPU is to speed up inference, this expection is not satisfied.
> 
> Token generation is limited by memory bandwidth, so you'll see very similar speeds for CPU or iGPU. Th iGPU helps with input token processing or image processing. I'm getting x2 to x3 speedups on input token processing or VLMs on an Intel 10th gen iGPU when using llama.cpp with Vulkan.

Do you mean iGPU can speed up Prefill phase but not Decode phase? @ddpasa

**Comment by MaoJianwei** [comment]
I found the solution! That's crazy! @NeoZhangJianyu 

https://github.com/ggml-org/llama.cpp/issues/1956

**Comment by vsenn** [comment]
+1. It would be great to have iGPU support for Intel hardware, as I have stuff like Intel UHD Graphics @ 1.10 GHz [Integrated], Intel Comet Lake UHD Graphics @ 1.15 GHz [Integrated] with 65 GB RAM\VRAM in my Intel NUC machines running Kubernetes cluster.

**Comment by MaoJianwei** [comment]
I found the solution! https://github.com/ggml-org/llama.cpp/issues/1956#issuecomment-3172333543
@vsenn

**Comment by escherstair** [comment]
I don't understand if with the trick from [ggml-org/llama.cpp#1956 (comment)](https://github.com/ggml-org/llama.cpp/issues/1956#issuecomment-3172333543) ollam supports Intel iRISx out of the box or not.

**Comment by MaoJianwei** [comment]
> I don't understand if with the trick from [ggml-org/llama.cpp#1956 (comment)](https://github.com/ggml-org/llama.cpp/issues/1956#issuecomment-3172333543) ollam supports Intel iRISx out of the box or not.

I suggest you use llama.cpp directly, not ollama
